### PR TITLE
Register array intrinsics in fortran backend

### PR DESCRIPTION
**Pull request by HarshitaKalani**
Towards: #4017 #420

### CODE REVIEW

**Comment by HarshitaKalani** [comment]
There can be a better and cleaner approach for this. 
We can create a function which takes the number of mandatory arguments and optional arguments along with the intrinsic id which will reduce the redundancy.
LMK if this sounds good, will proceed with that in that case.

**Review by certik — APPROVED** [review]

**Comment by certik** [comment]
I think your design idea is fine. Just keep in mind that in ASR we are not tied to Fortran and can (will) change the intrinsics design as we see fit. But currently there is a large overlap with Fortran of course. You can create the function you described and we can call it for intrinsics that behave exactly as in Fortran. And for the rest we'll do manual translation as we do now.

File: integration_tests/CMakeLists.txt [hunk 1]
```diff
@@ -688,14 +688,14 @@ RUN(NAME intrinsics_138 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran) #
...
+RUN(NAME intrinsics_142 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran) # dot_product
 RUN(NAME intrinsics_143 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # dot_product
-RUN(NAME intrinsics_144 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # dot_product
+RUN(NAME intrinsics_144 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran) # dot_product
 RUN(NAME intrinsics_145 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # dot_product
 RUN(NAME intrinsics_146 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # dprod
 RUN(NAME intrinsics_147 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # pack
 RUN(NAME intrinsics_148 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc) # pack
-RUN(NAME intrinsics_149 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc NOFAST) # unpack
+RUN(NAME intrinsics_149 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc NOFAST fortran) # unpack
```
Comment (Shaikh-Ubaid) [inline integration_tests/CMakeLists.txt:698]: I think this does not get considered as a part of LABELS at the moment.

```suggestion
RUN(NAME intrinsics_149 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc fortran NOFAST) # unpack
```

File: integration_tests/CMakeLists.txt [hunk 2]
```diff
@@ -1484,7 +1484,7 @@ RUN(NAME precision_02 LABELS gfortran) # TODO fix this test
 RUN(NAME array_unbounded_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc llvmStackArray)
 RUN(NAME array_unbounded_02 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc llvmStackArray)
 
-RUN(NAME matmul_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc c llvm_nopragma c_nopragma NOFAST)
+RUN(NAME matmul_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc c llvm_nopragma c_nopragma NOFAST fortran)
```
Comment (Shaikh-Ubaid) [inline integration_tests/CMakeLists.txt:1487]: Similarly

File: integration_tests/CMakeLists.txt (hunk 1 above)
Comment (HarshitaKalani) [reply integration_tests/CMakeLists.txt:698]: What was the issue here? Is it the sequence of labels?

Comment (Shaikh-Ubaid) [reply integration_tests/CMakeLists.txt:698]: Can you check if `fortran` is considered as a part of `LABELS` with your current change?

Comment (HarshitaKalani) [reply integration_tests/CMakeLists.txt:698]: Oh I see
When I run the following command:
```console
$ ./integration_tests/run_tests.py -b fortran
```
A number of tests run but intrinsics_149 is not present there.

Comment (HarshitaKalani) [reply integration_tests/CMakeLists.txt:698]: And it fails after applying the suggestion.
```console
CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o.tmp.f90:55:132:

   55 | print *, (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [1, 0, 1, 0], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
      |                                                                                                                                    1
Error: Line truncated at (1) [-Werror=line-truncation]
CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o.tmp.f90:55:132:

   55 | print *, (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [1, 0, 1, 0], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
      |                                                                                                                                    1
Error: Syntax error in array constructor at (1)
CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o.tmp.f90:56:132:

   56 | unity__ = (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [1, 0, 1, 0], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
      |                                                                                                                                    1
Error: Line truncated at (1) [-Werror=line-truncation]
CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o.tmp.f90:56:132:

   56 | unity__ = (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [1, 0, 1, 0], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
      |                                                                                                                                    1
Error: Syntax error in array constructor at (1)
CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o.tmp.f90:25:7:

   25 | mask = reshape([1, 0, 0, 1], shape(mask))
      |       1
Warning: Extension: Conversion from INTEGER(4) to LOGICAL(4) at (1)
CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o.tmp.f90:39:9:

   39 | mask_r = reshape([1, 0, 0, 1, 1, 0], shape(mask_r))
      |         1
Warning: Extension: Conversion from INTEGER(4) to LOGICAL(4) at (1)
CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o.tmp.f90:63:8:

   63 | mask_ = [1, 0, 1, 0]
      |        1
Warning: Extension: Conversion from INTEGER(4) to LOGICAL(4) at (1)
CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o.tmp.f90:72:9:

   72 | mask_c = reshape([1, 0, 1, 0, 1, 0, 1, 0], shape(mask_c))
      |         1
Warning: Extension: Conversion from INTEGER(4) to LOGICAL(4) at (1)
f951: some warnings being treated as errors
The command 'gfortran -fno-backtrace -o CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o -c CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o.tmp.f90' failed.
CMakeFiles/intrinsics_200.dir/intrinsics_200.f90.o.tmp.f90:6:7:

    6 | mask = [1, 0, 1, 0, 1]
      |       1
Warning: Extension: Conversion from INTEGER(4) to LOGICAL(4) at (1)
make[2]: *** [CMakeFiles/intrinsics_149.dir/build.make:75: CMakeFiles/intrinsics_149.dir/intrinsics_149.f90.o] Error 11
make[1]: *** [CMakeFiles/Makefile2:1983: CMakeFiles/intrinsics_149.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 60%] Linking Fortran executable intrinsics_174
[ 60%] Linking Fortran executable intrinsics_199
[ 60%] Linking Fortran executable intrinsics_175
[ 61%] Linking Fortran executable intrinsics_206
[ 61%] Linking Fortran executable intrinsics_200
[ 61%] Linking Fortran executable modules_06
[ 61%] Linking Fortran executable modules_01
[ 61%] Built target intrinsics_175
[ 61%] Built target intrinsics_174
[ 61%] Built target intrinsics_206
[ 61%] Built target intrinsics_199
[ 61%] Built target intrinsics_200
[ 61%] Built target modules_06
[ 61%] Built target modules_01
make: *** [Makefile:101: all] Error 2
Command failed.
```

Comment (Shaikh-Ubaid) [reply integration_tests/CMakeLists.txt:698]: Good, it means:

- `fortran` is not being considered as part of LABELS. Hence my suggestion above.
- this specific test case is not/yet-to-be supported by the `fortran` backend.

Comment (HarshitaKalani) [reply integration_tests/CMakeLists.txt:698]: Got it. Thanks for this, really helpful :)

**Comment by HarshitaKalani** [comment]
```console
$ lfortran integration_tests/intrinsics_149.f90 --show-fortran
program intrinsics_149
implicit none
integer(4), dimension(2, 2) :: field
real(4), dimension(4) :: field_
complex(4), dimension(2, 4) :: field_c
real(8), dimension(6) :: field_d
real(4), dimension(3, 2) :: field_r
logical(4), dimension(2, 2) :: mask
logical(4), dimension(4) :: mask_
logical(4), dimension(2, 4) :: mask_c
logical(4), dimension(6) :: mask_d
logical(4), dimension(3, 2) :: mask_r
integer(4), dimension(2, 2) :: unity
real(4), dimension(4) :: unity_
real(4), dimension(4) :: unity__
complex(4), dimension(2, 4) :: unity_c
real(8), dimension(6) :: unity_d
real(4), dimension(3, 2) :: unity_r
integer(4), dimension(2), save :: vector = [1, 1]
real(4), dimension(4) :: vector_
complex(4), dimension(8) :: vector_c
real(8), dimension(6) :: vector_d
real(4), dimension(6) :: vector_r
field = 0
mask = reshape([1, 0, 0, 1], shape(mask))
unity = (unpack(vector, mask, field))
print *, unity
if (sum(unity) /= 2) then
    error stop
end if
if (unity(1, 1) /= 1) then
    error stop
end if
if (unity(2, 2) /= 1) then
    error stop
end if
vector_r = 23.12000000
field_r = 681.31000000
mask_r = reshape([1, 0, 0, 1, 1, 0], shape(mask_r))
unity_r = (unpack(vector_r, mask_r, field_r))
print *, unity_r
print *, sum(unity_r)
if (abs(sum(unity_r) - 2113.29004000) > 0.00000001) then
    error stop
end if
vector_d = 23.1200000000000010d0
mask_d = .true.
field_r = real(681.3099999999999454d0, kind=4)
unity_d = (unpack(vector_d, mask_d, field_d))
print *, unity_d
print *, sum(unity_d)
if (abs(sum(unity_d) - 138.7199999999999989d0) > real(0.00000000, kind=8)) then
    error stop
end if
print *, (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [1, 0, 1, 0], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
unity__ = (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [1, 0, 1, 0], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
print *, unity__
print *, sum(unity__)
if (abs(sum(unity__) - 52.24000170) > 0.00000001) then
    error stop
end if
vector_ = [23.12000084, 23.12000084, 23.12000084, 23.12000084]
mask_ = [1, 0, 1, 0]
field_ = [1.00000000, 2.00000000, 3.00000000, 4.00000000]
unity_ = (unpack(vector_, mask_, field_))
print *, unity_
print *, sum(unity_)
if (abs(sum(unity_) - 52.24000170) > 0.00000001) then
    error stop
end if
vector_c = (23.12000000, -23.12000000)
mask_c = reshape([1, 0, 1, 0, 1, 0, 1, 0], shape(mask_c))
field_c = (1.10000000, -91.24000000)
unity_c = (unpack(vector_c, mask_c, field_c))
print *, unity_c
print *, abs(sum(unity_c))
if (abs(abs(sum(unity_c)) - 467.58642600) > 0.00000001) then
    error stop
end if
end program intrinsics_149
```
The above is the generated fortran code which on execution throws the following error:
```console
$ lfortran integration_tests/intrinsics_149.f90 
semantic error: The argument `mask` in `unpack` must be of type Logical
   --> integration_tests/intrinsics_149.f90:127:76
    |
127 |     print *, (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [1, 0, 1, 0], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
    |                                                                            ^^^^^^^^^^^^ 


Note: Please report unclear, confusing or incorrect messages as bugs at
https://github.com/lfortran/lfortran/issues.
```
which is rectified on changing the integers `1`, `0` to `.true.` and `.false.` respectively for mask argument.
```diff
program intrinsics_149
    implicit none
    integer(4), dimension(2, 2) :: field
    real(4), dimension(4) :: field_
    complex(4), dimension(2, 4) :: field_c
    real(8), dimension(6) :: field_d
    real(4), dimension(3, 2) :: field_r
    logical(4), dimension(2, 2) :: mask
    logical(4), dimension(4) :: mask_
    logical(4), dimension(2, 4) :: mask_c
    logical(4), dimension(6) :: mask_d
    logical(4), dimension(3, 2) :: mask_r
    integer(4), dimension(2, 2) :: unity
    real(4), dimension(4) :: unity_
    real(4), dimension(4) :: unity__
    complex(4), dimension(2, 4) :: unity_c
    real(8), dimension(6) :: unity_d
    real(4), dimension(3, 2) :: unity_r
    integer(4), dimension(2), save :: vector = [1, 1]
    real(4), dimension(4) :: vector_
    complex(4), dimension(8) :: vector_c
    real(8), dimension(6) :: vector_d
    real(4), dimension(6) :: vector_r
    field = 0
    mask = reshape([1, 0, 0, 1], shape(mask))
    unity = (unpack(vector, mask, field))
    print *, unity
    if (sum(unity) /= 2) then
        error stop
    end if
    if (unity(1, 1) /= 1) then
        error stop
    end if
    if (unity(2, 2) /= 1) then
        error stop
    end if
    vector_r = 23.12000000
    field_r = 681.31000000
    mask_r = reshape([1, 0, 0, 1, 1, 0], shape(mask_r))
    unity_r = (unpack(vector_r, mask_r, field_r))
    print *, unity_r
    print *, sum(unity_r)
    if (abs(sum(unity_r) - 2113.29004000) > 0.00000001) then
        error stop
    end if
    vector_d = 23.1200000000000010d0
    mask_d = .true.
    field_r = real(681.3099999999999454d0, kind=4)
    unity_d = (unpack(vector_d, mask_d, field_d))
    print *, unity_d
    print *, sum(unity_d)
    if (abs(sum(unity_d) - 138.7199999999999989d0) > real(0.00000000, kind=8)) then
        error stop
    end if
- print *, (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [.1, 0, 1, 0], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
+ print *, (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [.true., .false., .true., .false.], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
- unity__ = (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [1, 0, 1, 0], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
+ unity__ = (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [.true., .false., .true., .false.], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
    print *, unity__
    print *, sum(unity__)
    if (abs(sum(unity__) - 52.24000170) > 0.00000001) then
        error stop
    end if
    vector_ = [23.12000084, 23.12000084, 23.12000084, 23.12000084]
- mask_ = [1, 0, 1, 0]
+ mask_ = [.true., .false., .true., .false.]
    field_ = [1.00000000, 2.00000000, 3.00000000, 4.00000000]
    unity_ = (unpack(vector_, mask_, field_))
    print *, unity_
    print *, sum(unity_)
    if (abs(sum(unity_) - 52.24000170) > 0.00000001) then
        error stop
    end if
    vector_c = (23.12000000, -23.12000000)
- mask_c = reshape([1, 0, 1, 0, 1, 0, 1, 0], shape(mask_c))
+ mask_c = reshape([.true., .false., .true., .false., .true., .false., .true., .false.], shape(mask_c))
    field_c = (1.10000000, -91.24000000)
    unity_c = (unpack(vector_c, mask_c, field_c))
    print *, unity_c
    print *, abs(sum(unity_c))
    if (abs(abs(sum(unity_c)) - 467.58642600) > 0.00000001) then
        error stop
    end if
end program intrinsics_149
```

**Comment by HarshitaKalani** [comment]
Solved the above but realised the the actual issue is with truncation of line.
```console
$ gfortran a.f90 && ./a.out
a.f90:55:132:

   55 |     print *, (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [.true., .false., .true., .false.], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
      |                                                                                                                                    1
Error: Line truncated at (1) [-Werror=line-truncation]
a.f90:55:132:

   55 |     print *, (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [.true., .false., .true., .false.], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
      |                                                                                                                                    1
Error: Syntax error in array constructor at (1)
a.f90:56:132:

   56 |     unity__ = (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [.true., .false., .true., .false.], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
      |                                                                                                                                    1
Error: Line truncated at (1) [-Werror=line-truncation]
a.f90:56:132:

   56 |     unity__ = (unpack([23.12000084, 23.12000084, 23.12000084, 23.12000084], [.true., .false., .true., .false.], [1.00000000, 2.00000000, 3.00000000, 4.00000000]))
      |                                                                                                                                    1
Error: Syntax error in array constructor at (1)
f951: some warnings being treated as errors
```

**Pull request merged**